File: Src/Runtime/Dictionaries/Size.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/25/26.
//

#include "Size.h"

#include "ADT/Lang/AST.h"
#include "ADT/Runtime/Object.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::Dictionaries;

ADT::Runtime::Object Dictionaries::Size(
    ADT::List::Object &args,
    ADT::Lang::AST *trace
)
{
    auto &dict = Support::Runtime::GetDictionaryObj(args[0]);

    return {
        ADT::Runtime::ObjectType::Integer,
        (int)dict->size()
    };
}

File: src/runtime/dictionaries/size.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use crate::adt::lang::{RuntimeArguments, AST};
use crate::adt::result::ExecutionTupleResult;
use crate::adt::runtime::{HashableObject, Object};
use crate::support::runtime::execution::continue_execution;
use crate::support::runtime::object::get_dictionary;

pub fn size<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionTupleResult<'a>
{
    let dict_obj = args.get(0).unwrap();
    let dict = get_dictionary(dict_obj, trace)?;
    let dict = dict.borrow();

    continue_execution(
        Object::Hashable(
            HashableObject::Integer(
                dict.len() as isize
            )
        )
    )
}